test/metabase/query_processor_test/explicit_joins_test.clj

```diff
@@ -983,3 +989,40 @@
             (is (= [["2016-05-01T00:00:00Z" 3 nil nil]
                     ["2016-06-01T00:00:00Z" 2 "2016-06-01T00:00:00Z" 1]]
                    (mt/rows (qp/process-query query))))))))))
+
+(deftest test-31769
+  (testing "Make sure queries built with MLv2 that have source Cards with joins work correctly (#31769) (#33083)"
+    (mt/dataset sample-dataset
+      (let [metadata-provider (lib.metadata.jvm/application-database-metadata-provider (mt/id))]
+        (t2.with-temp/with-temp [:model/Card {card-1-id :id} {:dataset_query
+                                                              (lib.convert/->legacy-MBQL
+                                                               (lib.tu.mocks-31769/card-1-query metadata-provider mt/id))}
+                                 :model/Card {card-2-id :id} {:dataset_query
+                                                              (lib.convert/->legacy-MBQL
+                                                               (lib.tu.mocks-31769/card-2-query metadata-provider mt/id))}]
+          (testing "returned columns"
+            (let [card (lib.metadata/card metadata-provider card-1-id)]
+              (assert card)
+              (testing "with Card query as previous stage"
+                (let [query (-> (lib/query metadata-provider (:dataset-query card))
+                                lib/append-stage)
+                      cols  (lib/returned-columns query)]
+                  (is (=? [[:field {:base-type :type/Text} "Products__CATEGORY"]
+                           [:field {:base-type :type/Integer} "count"]]
+                          (map lib.ref/ref cols)))))
+              (testing "with :source-card"
+                (let [query (lib/query metadata-provider card)
+                      cols  (lib/returned-columns query)]
+                  (is (=? [[:field {:base-type :type/Text} "Products__CATEGORY"]
+                           [:field {:base-type :type/Integer} "count"]]
+                          (map lib.ref/ref cols)))))))
+          (let [query (lib.convert/->legacy-MBQL
+                       (lib.tu.mocks-31769/query metadata-provider card-1-id card-2-id))]
+            (is (=? {:query {:joins [{:condition [:=
+                                                  [:field "Products__CATEGORY" {:base-type :type/Text}]
+                                                  [:field integer? {:base-type :type/Text, :join-alias string?}]]}]}}
```

Comment:
Why do we have an int reference here?

Reply:
Because fixing https://github.com/metabase/metabase/issues/29763 is out of scope for this PR

Reply:
👍 Please add a comment with the reference to #29763. It's already difficult to tell which parts of the tests are _really_ meant as specification.